### ISSUE TITLE
:term flicker on each keypress

### ISSUE THREAD

**Issue by sbalnave** [neovim/neovim#11915]
- `nvim --0.5.0-2538e6151
- Operating system/version: OSX 10.15
- Terminal name/version: Kitty
- `$TERM`: screen-256color

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
:set scrolloff=5
:term
ls -l
ls -l
(i.e. whole terminal screen full)
clear  (or vim)
start typing (or enter insert mode and start typing)
```

### Actual behaviour

with every keypress the screen flickers momentarily showing the last 5 lines of the terminal session at the top of the window

### Expected behaviour

should work without flicker just like it does with scroll off set to 0

**Comment by KillTheMule** [comment]
Reproduced on linux with `NVIM v0.5.0-330-g71ee46acc` on gnome-terminal, konsole, xterm, st (if needed, I can look up the version, but I guess it doesn't matter at thispoint).

**Comment by lacygoill** [comment]
Possible duplicate of #11072.

FWIW, I had the same issue whenever I used to expand a zsh snippet; I fixed it by temporarily resetting `'scrolloff'` to 0 in a terminal buffer.

As a workaround, you could use this:

    autocmd TermOpen,TermEnter * setlocal scrolloff=0
    autocmd TermLeave * setlocal scrolloff=5

Although, there is still one issue; `'scrolloff'` is a global option in Nvim.  But it should become window-local once the PR #11854 is merged.

**Comment by WillEccles** [comment]
Working on #11854 now. Need to make the linter not hate me, but the functionality is there.

**Comment by WillEccles** [comment]
Note also that I can confirm that after #11854, the solution of using `setl so=0` on `TermOpen` works to circumvent this issue. Using macOS 10.14.4 and Alacritty.

**Comment by bfredl** [comment]
On top of #11854 my suggestion https://github.com/neovim/neovim/issues/11072#issuecomment-533828802 should also be easy to implement:  `get_scrolloff_value` should return 0 when `State & TERM_FOCUS`.

**Comment by WillEccles** [comment]
I could add this to #11854's changes, but this would be a little outside the scope of the vim patch.

**Comment by bfredl** [comment]
sure it could be a follow-up PR

**Comment by sbalnave** [comment]
https://github.com/neovim/neovim/issues/11915#issuecomment-589068664 works for me, but the :term window jumps down 5 lines when switching to another split window, so yes a fix would need scrolloff to be window-local

**Comment by lacygoill** [comment]
https://github.com/neovim/neovim/commit/87d892afa0475644e91d9c8a57b7c35491c4dc32 just got merged.
